modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DBBasedExpiryMessageDeletionTask.java

```diff
@@ -0,0 +1,169 @@
+import org.wso2.andes.store.HealthAwareStore;
+import org.wso2.andes.store.StoreHealthListener;
+
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This task is responsible to delete the expired messages from the database which were not currently
+ * allocated to any slots. If we delete any messages that are assigned to a slot delivery worker, the
+ * content may got deleted when the content is asked to retrieve. This can mess up the message delivery
+ * flow. Because of that the messages not allocated to a slot are considered as safe to delete
+ */
+public class DBBasedExpiryMessageDeletionTask implements Runnable, StoreHealthListener {
+
+    private static Log log = LogFactory.getLog(SetBasedExpiryMessageDeletionTask.class);
+
+    /**
+     * Indicates and provides a barrier if messages stores become offline.
+     * marked as volatile since this value could be set from a different thread
+     */
+    private volatile SettableFuture<Boolean> messageStoresUnavailable;
+
+    /**
+     * Used to perform database operations on the context store.
+     */
+    private AndesContextStore andesContextStore = AndesContext.getInstance().getAndesContextStore();
+
+    /**
+     *  The slot gap need to be maintained as not reachable for deletion task as they may be allocated
+     *  to a slot delivery worker in the near future
+     */
+    private final int safetySlotCount;
+
+
+    public DBBasedExpiryMessageDeletionTask() {
+
+        this.messageStoresUnavailable = null;
+        // Register AndesRecoveryTask class as a StoreHealthListener
+        FailureObservingStoreManager.registerStoreHealthListener(this);
+
+        safetySlotCount = AndesConfigurationManager.readValue
+                (AndesConfiguration.PERFORMANCE_TUNING_SAFE_DELETE_REGION_SLOT_COUNT);
+
+    }
+
+    @Override
+    public void run() {
+
+        boolean isClusterEnabled = AndesContext.getInstance().isClusteringEnabled();
+
+        try {
+            /**
+             * This logic belongs to an MB run in stand alone mode / the coordinator node run in cluster mode
+             */
+            if (!isClusterEnabled || (isClusterEnabled && AndesContext.getInstance().getClusterAgent().isCoordinator())) {
+
+                Set<String> queues = andesContextStore.getAllQueues();
+
+                for (String queueName : queues) {
+                    //get the upper bound messageID for each unassigned slots as a set for the specific queue
+                    TreeSet<Long> messageIDSet = andesContextStore.getMessageIds(queueName);
+
+                    /**
+                     * deletion task run only if there are more messages than the safety slot count upper bound
+                     * otherwise deletion task is not run
+                     * minimum safetySlotCount is 1
+                     */
+                    if (safetySlotCount >= 1 && messageIDSet.size() >= safetySlotCount) {
```

Comment:
safetySlotCount >= is a check for a valid config value is provided or not. we don't need that in the logic. shall we validate the configuration during the server start up?